### PR TITLE
GH-40333: [Docs] Improve env var docs for ARROW_USER_SIMD_LEVEL

### PR DESCRIPTION
**Pull request by amoeba**
### Rationale for this change

Conversation in https://github.com/apache/arrow/issues/40333.

### What changes are included in this PR?

Just tweaks to the text in docs/source/cpp/env_vars.rst.

### Are these changes tested?

I rendered them locally.

### Are there any user-facing changes?

Just docs here.
* GitHub Issue: #40333

### CODE REVIEW

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

File: docs/source/cpp/env_vars.rst
Comment (wjones127) [inline docs/source/cpp/env_vars.rst:160]: For clarity, maybe we should share the mechanism. Also, I think it's more appropriate to call this the *minimum* SIMD level, as it is older computers that won't support this.
```suggestion
      The ``ARROW_SIMD_LEVEL`` variable sets the minimum supported SIMD level at the
      compiler level. This works by passing the appropriate target architecture flag, such as ``-march=haswell`` for AVX2, to the compiler.  Unlike runtime dispatch, compile-time SIMD optimizations
```

https://github.com/apache/arrow/blob/3d467ac7bfae03cf2db09807054c5672e1959aec/cpp/cmake_modules/SetupCxxFlags.cmake#L47-L67

File: docs/source/cpp/env_vars.rst
Comment (wjones127) [inline docs/source/cpp/env_vars.rst:163]:
```suggestion
      cannot be changed at runtime. For example, if you compile Arrow C++ with
      AVX512 enabled, the resulting binary will only run on AVX512-enabled
      CPUs.  Setting ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of
```

**Review by wjones127 — COMMENTED** [review]
Thanks for writing this up. I have a couple minor suggestions.

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst]: I like this change, thanks.

**Comment by amoeba** [comment]
Thanks @wjones127, both changes accepted as they were good improvements.

**Review by wjones127 — APPROVED** [review]
Looks good. Though you might want to fix the line length my suggestion broke. 🤦

**Comment by amoeba** [comment]
Whoops. Good eye. Done and pushed.

**Comment by zanmato1984** [comment]
+1

The improvement is very nice and clear.

File: docs/source/cpp/env_vars.rst
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:160]: This is a bit misleading IMHO. It sets which instructions the compiler is allowed to emit regardless of dynamic dispatch.

File: docs/source/cpp/env_vars.rst
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:171]: PyArrow shouldn't really be mentioned here, these are the C++ docs.

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:160]: Ah, I think I see now. I just took a closer look at how ARROW_SIMD_LEVEL gets handled and the options don't stack onto each other. Borrowing from your language, how's this?

```suggestion
      The ``ARROW_SIMD_LEVEL`` variable controls which SIMD instructions the compiler is allowed to emit. 
 This works by passing the appropriate target
```

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:171]:
```suggestion
      Instruction).  In this case, one must rebuild Arrow from
```

Comment (amoeba) [reply docs/source/cpp/env_vars.rst:171]: Removed the mention. Thanks.

**Comment by amoeba** [comment]
Thanks for the review @pitrou. I proposed two suggestions, let me know if those look good to you.

File: docs/source/cpp/env_vars.rst
Comment (mapleFU) [inline docs/source/cpp/env_vars.rst:138]: Just curious, what does `buggy` here means? ( I'm not familiar with SIMD ). Would that's the case that the arch don't implement the instr well? Or just don't have this instr, do user need disable it?

File: docs/source/cpp/env_vars.rst
Comment (mapleFU) [inline docs/source/cpp/env_vars.rst:168]: Hmmm 🤔 A bit misleading here, since we've some hand-written SSE4.2 code, and some compiler generated code using some auto-vectorization techniques?

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:138]: Thanks for taking a look. There may be a better word here...By buggy I was referring to the comment in https://github.com/apache/arrow/issues/40333#issuecomment-1977142577 mentioning situations where like running under a hypervisor where cpu detection reports that an instruction is supported but it's not handled correctly by the hypervisor. I'm not familiar with this so I don't know what "not handled correctly" really means.

Maybe instead of buggy, say "where instructions are supported but are not performant or cause other issues"?

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:168]: I'm not familiar with either of those bits of the codebase but I can take a look tomorrow. Can you drop some links if you have them handy? (Logging off for the night here)

File: docs/source/cpp/env_vars.rst
Comment (mapleFU) [reply docs/source/cpp/env_vars.rst:138]: Looks nice, we can wait for @pitrou and others' idea here

File: docs/source/cpp/env_vars.rst
Comment (mapleFU) [reply docs/source/cpp/env_vars.rst:168]: https://github.com/apache/arrow/issues/40333#issuecomment-1976015427

I've pasted some in issue

Comment (amoeba) [reply docs/source/cpp/env_vars.rst:168]: Thanks. It looks like, in those cases, the symbols get compiled in but, at least for the example in level_comparison.cc, the code looks like it runs through the dispatch machinery in dispatch.h so it still wouldn't be selected if `ARROW_USER_SIMD_LEVEL` is set in such a way that prevents it. All that code is new to me so I'd appreciate another set of eyes.

File: docs/source/cpp/env_vars.rst
Comment (pitrou) [reply docs/source/cpp/env_vars.rst:138]: I agree with "where instructions are supported but are not performant or cause other issues"?

File: docs/source/cpp/env_vars.rst
Comment (pitrou) [reply docs/source/cpp/env_vars.rst:160]: Hmm, this is still misleading in a different way. SIMD instructions are also emitted for runtime-selected optimizaitions, and they are not controlled by `ARROW_SIMD_LEVEL` :-)

Perhaps my suggestion was a dead end afterwards, and we shouldn't try to phrase this in terms of "emit[ting] SIMD instructions"?

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst]: Thanks! Change made in 5ad7282f7bc3581825098d3789fa0af0c3f0af5f.

File: docs/source/cpp/env_vars.rst
Comment (pitrou) [inline docs/source/cpp/env_vars.rst:155]: Ok, perhaps instead of making this note about CMake configuration variables, we should explain the two different kinds of SIMD optimizations? This would be more user-oriented (this is a page about environment variables after all).

Something like:
> In addition to runtime-selected SIMD optimizations dispatch, Arrow C++ can also be compiled
> with SIMD optimizations that are cannot be disabled at runtime. For example, by default, SSE4.2
> optimizations are enabled on x86 builds: therefore, with this default setting, Arrow C++ does
> not work at all on a CPU without support for SSE4.2. This setting can be changed using
> the``ARROW_SIMD_LEVEL`` CMake variable so as to either raise or lower the optimization level.
>
> Finally, the ``ARROW_RUNTIME_SIMD_LEVEL`` CMake variable sets a compile-time upper bound
> to runtime-selected SIMD optimizations. This useful in cases where a compiler reports support
> for an instruction set but does not actually support it in full.

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:168]: No worries, and thank you for the help with the text here @mapleFU. It's appreciated.

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst]: You make a good point about the approach. I like what you wrote and incorporated it in 46bcac8bec4896cec739ebb1151e82f1e54362f2. It's also shorter which is a plus.

**Comment by amoeba** [comment]
Thanks for taking another look at this @pitrou. I think the new paragraph you added addresses both of your most recent comments. With that, I think this is very close if not ready to be merged.

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

**Review by mapleFU — APPROVED** [review]
Current doc LGTM

File: docs/source/cpp/env_vars.rst
Comment (zanmato1984) [inline docs/source/cpp/env_vars.rst:156]:
```suggestion
      also be compiled with SIMD optimizations that cannot be disabled at
```

**Review by zanmato1984 — COMMENTED** [review]
Found a typo.

**Review by pitrou — APPROVED** [review]

File: docs/source/cpp/env_vars.rst
Comment (amoeba) [reply docs/source/cpp/env_vars.rst:156]: Thank you!

Comment (amoeba) [reply docs/source/cpp/env_vars.rst:156]: Sorry, got distracted by a phone call. Will push up in a sec.

Comment (amoeba) [reply docs/source/cpp/env_vars.rst]: See 0aab71c1e3f87218571aaca6bcb09d23df37fb87.

**Review by zanmato1984 — APPROVED** [review]
+1

**Comment by mapleFU** [comment]
Will wait for some days, and would merge it if no negative comments

**Comment by mapleFU** [comment]
Thanks all, merged!

**Pull request merged**
